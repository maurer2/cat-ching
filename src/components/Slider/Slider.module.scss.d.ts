import globalClassNames, { ClassNames as GlobalClassNames } from '../..style.d';
declare const classNames: typeof globalClassNames & {
  readonly slider: 'slider';
  readonly slide: 'slide';
};
export default classNames;
export type ClassNames = 'slider' | 'slide' | GlobalClassNames;
