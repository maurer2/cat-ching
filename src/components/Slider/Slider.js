import React from 'react';
import style from './Slider.module.scss';

function Slider(props) {
  return (
    <ul className={style.slider}>
      {props.children.map((child, index) => (
        <li className={style.slide} key={index}>
          {child}
        </li>
      ))}
    </ul>
  );
}

export default Slider;
