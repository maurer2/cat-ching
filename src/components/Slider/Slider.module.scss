.slider {
  display: flex;
  flex-grow: 1;
  margin: 0;
  padding: 0;
  list-style: none;
  overflow-y: scroll;
  scroll-snap-type: x mandatory;
}

.slide {
  display: flex;
  min-width: 100%;
  justify-content: center;
  align-items: center;
  scroll-snap-align: center;

  &:nth-child(odd) {
    background: whitesmoke;
  }
}
