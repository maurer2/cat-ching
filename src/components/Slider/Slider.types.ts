export type SliderProps = Record<string, never>;
