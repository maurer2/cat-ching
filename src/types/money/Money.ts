/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/lines-between-class-members */
export const currencies = {
  GBP: 'Â£',
  USD: '$',
} as const;

export type CurrencyName = keyof typeof currencies;

export default class Money {
  private readonly _value: bigint;
  private readonly _name: CurrencyName;
  private readonly _formatter: Intl.NumberFormat;

  public static fromBigInt(
    valueInFractions: bigint,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    return new this(valueInFractions, name, formatter);
  }

  public static fromNumber(
    valueInFractions: number,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    const valueAsBigInt = BigInt(valueInFractions);

    return new this(valueAsBigInt, name, formatter);
  }

  public static fromRandom(name: CurrencyName, formatter?: Intl.NumberFormat): Money {
    const randomNumber: number = Math.ceil(Math.random() * (1000 - 1)) + 1;
    const valueAsBigInt = BigInt(randomNumber);

    return new this(valueAsBigInt, name, formatter);
  }

  private constructor(valueInFractions: bigint, name: CurrencyName, formatter?: Intl.NumberFormat) {
    this._value = valueInFractions;
    this._name = name;
    this._formatter = formatter
      ?? new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency: 'GBP',
        useGrouping: false, // disable thousand separator
      });
  }

  public add(valueToAdd: Money): Money {
    const newValue: bigint = this._value + valueToAdd._value;

    return new Money(newValue, this._name, this._formatter);
  }

  public isEqualTo(secondValue: Money): boolean {
    return this._value === secondValue._value;
  }

  public isNegative(): boolean {
    return this._value < 0;
  }

  public get valueAsSeparateParts(): Intl.NumberFormatPart[] {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.formatToParts(valueAsNumber);
  }

  public get valueAsFormattedString(): string {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.format(valueAsNumber);
  }

  public get valueFormattedAsIntegerAndFraction(): string {
    // https://stackoverflow.com/a/70086020
    const integer = Number(this._value / 100n);
    const fraction = Number(this._value % 100n);
    // const integral = Number(this.value / 100n);
    // const fraction = Number(this.value % 100n);

    return `${integer}.${`${fraction}`.padEnd(2, '0')}`;
  }
}

export type MoneyType = typeof Money;
