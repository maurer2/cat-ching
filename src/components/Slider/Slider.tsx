import React, { useRef } from 'react';

import style from './Slider.module.scss';
import * as Types from './Slider.types';

function Slider({ children }: Types.SliderProps): JSX.Element {
  const scrollWrapperElement = useRef<HTMLUListElement | null>(null);

  // function handleSegmentScrolling(event: WheelEvent<HTMLUListElement>) {
  //   const { deltaY, currentTarget } = event;
  //   const slideWidthBB = currentTarget.getBoundingClientRect();
  //   const startScrollPosition = scrollWrapperElement.current?.scrollLeft ?? 0;

  //   // scroll right
  //   if (Math.sign(deltaY) === 1) {
  //     scrollWrapperElement.current?.scrollTo({
  //       left: startScrollPosition + slideWidthBB.width / 2,
  //       behavior: 'smooth',
  //     });
  //   }

  //   // scroll left
  //   if (Math.sign(deltaY) === -1) {
  //     scrollWrapperElement.current?.scrollTo({
  //       left: startScrollPosition - slideWidthBB.width / 2,
  //       behavior: 'smooth',
  //     });
  //   }
  // }

  return (
    <ul
      className={style.slider}
      ref={scrollWrapperElement}
      // onWheel={handleSegmentScrolling}
      data-testid="slider"
    >
      {React.Children.map(children, (child) => (
        <li className={style.slide}>{child}</li>
      ))}
    </ul>
  );
}

export default Slider;
