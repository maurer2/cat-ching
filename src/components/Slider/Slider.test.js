import React from 'react';
import { shallow } from 'enzyme';

import Slider, { Entry, Container } from './Slider';
import Money from '../../data/money';

describe('Slider', () => {
  /*
  const mockeHandleAmountChange = jest.fn();
  const moneyData = new Money(1, 'Dollar');
  const coinData = {
    name: '1 dollar',
    image: 'image name',
    amount: moneyData,
    handleAmountChange: mockeHandleAmountChange,
    key: '1 dollar',
  }
  */

  it('should render correctly', () => {
    expect(true).toBe(true);
  });
})

describe('Entry', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Entry>
      Test
    </Entry>);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Container', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Container>
      Test
    </Container>);

    expect(wrapper).toMatchSnapshot();
  });
});
