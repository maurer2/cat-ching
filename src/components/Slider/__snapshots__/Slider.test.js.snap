// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container should render correctly 1`] = `
<ul
  className="slider"
>
  Test
</ul>
`;

exports[`Entry should render correctly 1`] = `
<li
  className="slide"
>
  Test
</li>
`;
